src/main/java/kitchenpos/menus/tobe/domain/application/ChangeMenuPrice.java

```diff
@@ -0,0 +1,42 @@
+package kitchenpos.menus.tobe.domain.application;
+
+import java.math.BigDecimal;
+import java.util.NoSuchElementException;
+import java.util.UUID;
+import kitchenpos.menus.tobe.domain.entity.Menu;
+import kitchenpos.menus.tobe.domain.repository.MenuRepository;
+import kitchenpos.menus.tobe.domain.vo.MenuPrice;
+import kitchenpos.menus.tobe.dto.MenuChangePriceDto;
+import org.springframework.stereotype.Service;
+
+@FunctionalInterface
+public interface ChangeMenuPrice {
+    Menu execute(UUID menuId, MenuChangePriceDto menuChangePriceDto);
+}
+
+@Service
+class DefaultChangeMenuPrice implements ChangeMenuPrice {
+    private final MenuRepository menuRepository;
+    private final CalculateSumOfMultiplyingMenuProductPriceAndMenuProductQuantity calculateSumOfMultiplyingMenuProductPriceAndMenuProductQuantity;
+
+    public DefaultChangeMenuPrice(MenuRepository menuRepository,
+                                  CalculateSumOfMultiplyingMenuProductPriceAndMenuProductQuantity calculateSumOfMultiplyingMenuProductPriceAndMenuProductQuantity) {
+        this.menuRepository = menuRepository;
+        this.calculateSumOfMultiplyingMenuProductPriceAndMenuProductQuantity = calculateSumOfMultiplyingMenuProductPriceAndMenuProductQuantity;
+    }
+
+    @Override
+    public final Menu execute(UUID menuId, MenuChangePriceDto menuChangePriceDto) {
+        final MenuPrice menuPrice = MenuPrice.of(menuChangePriceDto.getPrice());
+        final Menu menu = menuRepository.findMenuById(menuId)
+                                        .orElseThrow(NoSuchElementException::new);
+        BigDecimal sum = calculateSumOfMultiplyingMenuProductPriceAndMenuProductQuantity.execute(menu);
+
+        if (menuPrice.getValue().compareTo(sum) > 0) {
+            throw new IllegalArgumentException();
+        }
+        menu.changePrice(menuPrice);
```

Comment:
만약 `MenuProduct`가 price 정보를 가진다면 아래와 같이 변경해볼 수 있을 것 같은데, MenuProduct가 price를 가지도록 변경해보면 어떨까요? 

```suggestion
        menu.changePrice(MenuPrice.of(menuChangePriceDto.getPrice()));
```

Reply:
넵 한 번 적용해봤습니다. 리뷰해주시면 감사하겠습니다.

다만 가격필드를 Product, MenuProduct  에 두 번 저장하는 것은 data inconsistency 가 발생할 가능성이 너무 높은 것 같습니다. 
만약 실제로 이렇게 한다면 상품 가격 변경할 때 모든 MenuProduct 의 가격을 Transactional 하게 업데이트하거나 event 처리를 해야할 것 같습니다. 이번 미션에서는 어그리게이트를 나누는걸 고려해보는 걸 하다보니 이벤트 처리로 진행했습니다.
개인적인 의견으로는 저는 가격이 eventual consistency 를 적용할만한 필드가 아니라고 생각합니다. 바뀌는 그 즉시 적용이 되야하는 필드라고 생각합니다.

Reply:
이런 `eventual consistency`를 적용할만한가 아닌가는 결국 이 서비스의 특성을 생각할 수 밖에 없는데요. (말씀주신것도 동일하게 서비스의 특성상 적용할만한 필드가 아니라고 말씀하신거라고 생각되네요. 🙂)

Product의 가격이 변경됐을 때, Menu의 MenuProduct 가격이 변경되는 이 시점이 길어지면 길어질수록 (심지어 찰나의 지연일지라도) sync가 되기 전까지는 숨김처리가 되어야하는데 되지않아서 그 사이에 메뉴가 더 판매될 수도 있겠네요. 🤔

메뉴를 판매하는 가게 주인의 입장에서는 상품 가격이 변경되서 숨김처리된 메뉴로 주문이 들어온다면 문제가될까요? 👀

실제로 메뉴가 판매되는 것은 Menu의 `MenuPrice` 가격이고 이 가격은 Product의 ProductPrice가 변경됐다고해서 변경되는 것은 아닙니다.  결제 금액이 달라진다면 이슈가 생길 수 있지만 단순히 메뉴 상품의 가격 정보가 변경되는 것이고 추가적으로 메뉴 상품들의 가격 합이 메뉴의 가격보다 작으면 숨김 메뉴가 된다는 정책이 실시간으로 적용되지 않더라도 저는 가게주인에게 영향이 없다고 생각됩니다. (오히려 상품 가격은 내려갔는데  더 비싸게 메뉴를 팔았으니 금액적으로만 생각하면 더 이득일 수도 있겠네요. 🤑) 

저는 이런 관점에서 지금의 정책에서는 menuProduct의 menuProductPrice가 무조건 같은 트랜잭션안에서(혹은 실시간)  처리될 필요는 없다고 생각됩니다. 🙂